src/v/pandaproxy/schema_registry/json.cc

```diff
@@ -1956,6 +1974,169 @@ void sort(json::Value& val) {
     }
 }
 
+auto is_already_validated(const jsoncons::jsonpointer::json_pointer& p) {
+    // root objects or objects under "definitions" are already validated. note
+    // that other keywords are validated, but this a heuristic that takes a
+    // conservative approach.
+    if (p.empty()) {
+        return true;
+    }
+
+    return std::ranges::any_of(
+      p, [](std::string_view token) { return token == "definitions"; });
+}
+
+void collect_bundled_schemas_and_fix_refs(
+  id_to_schema_pointer& bundled_schemas,
+  jsoncons::uri base_uri,
+  jsoncons::jsonpointer::json_pointer this_obj_ptr,
+  jsoncons::json& this_obj,
+  json_schema_dialect dialect) {
+    // scan the json schema object for bundled schemas. limit the search to
+    // schemas under "defitions" or "$defs", as these are validated by the
+    // metaschema. A bundled schema is defined as a schema with "$id" :
+    // a_base_uri. all relative refs under this bundled schema will be relative
+    // to a_base_uri. if this_obj is a bundled schema, it might use a different
+    // dialect than the root schema, and the current implementation requires
+    // that the dialect is known. the jsonschema spec is not strict so we could
+    // receive some valid but nonsensical edge cases. these are not supported
+    // but do not generate errors here, since they are problematic only when a
+    // accessed by a $ref. so we don't register them in the base_uris_map, an
+    // will raise an error later if we try to access them. keyword | dialect  |
+    // is bundled schema
+    //   "$id" |    any   |       yes
+    //   "$id" | >=draft6 |       yes
+    //   "$id" |  draft4  |       no
+    //   "id"  |    any   |       yes
+    //   "id"  | >=draft6 |       no
+    //   "id"  |  draft4  |       yes
+
+    // try to extract the dialect from the $schema keyword, or use the parent
+    // dialect. empty means that "$schema" is present but the dialect is not
+    // known, and we should stop scanning this branch.
+    auto maybe_new_dialect = [&]() -> std::optional<json_schema_dialect> {
+        auto dialect_it = this_obj.find("$schema");
+        if (dialect_it == this_obj.object_range().end()) {
+            // we can keep using the parent dialect
+            return dialect;
+        }
+
+        // we have a $schema keyword, use this dialect if we find out that
+        // this_obj is a bundled schema
+        return from_uri(dialect_it->value().as_string_view());
+    }();
+
+    auto maybe_draft4_id_it = this_obj.find("id");
+    auto maybe_id_it = this_obj.find("$id");
```

Comment:
i tried to come with a simpler form but the fact is that we know that we are in bundled schema if we find a "id" member, and then if we are in a bundled schema then "$schema" can be analyzed, and the dialect has agree to "$id" or "id. it's kinda a chicken/egg problem,
so that's why i just try to find both, and if at least one is found then i assume we are in bundled schema and i can check if it's correct or not.